### ISSUE TITLE
ofi_layered_prov_getinfo uninitialized warning

### ISSUE THREAD

**Issue by jdinan** [ofiwg/libfabric#2013]
This warning looks like a potential bug:

```
prov/util/src/util_attr.c: In function ‘ofi_layered_prov_getinfo’:
prov/util/src/util_attr.c:116: warning: ‘tail’ may be used uninitialized in this function
```

@a-ilango, could you please take a look?

**Comment by shefty** [comment]
This looks like a false positive to me.  Layer_info and tail are either both NULL or both set.

**Comment by a-ilango** [comment]
Yes. Also, somehow I don't get this warning in my build. We can init tail to NULL to be safe.

**Comment by shefty** [comment]
The warning comes from here:

	if (!layer_info)
		layer_info = temp;
	else
		tail->next = temp;  <<<<<<
	tail = temp;

So, initializing to NULL doesn't completely solve this as tail is dereferenced before being set.  I don't see a great solution, but we could set tail = NULL and modify the else clause to:

	else if (tail)

Backing up, renaming layer_info to head may help with code readability.  However, this segment looks suspect:

	if (get_base_info) {
		*info = base_info;
	} else {
		for (fi = base_info; fi; fi = fi->next) {
			ret = alter_base_info(fi, &temp);
			if (ret)
				goto err3;
			if (!layer_info)
				layer_info = temp;
			else
				tail->next = temp;
			tail = temp;
		}
		fi_freeinfo(base_info);
	}
	fi_freeinfo(base_hints);
	*info = layer_info; 

Note that in the outer if check, *info is set to base_info.  That gets overwritten later.

**Comment by a-ilango** [comment]
Yes, that's a bug. :) Setting of *info should be inside the else clause.

**Comment by shefty** [comment]
I have a local fix

**Comment by a-ilango** [comment]
I was about to work on this but if you already have a fix please go ahead. Thanks!

**Comment by shefty** [comment]
Should be 'fixed' with PR #2014
